### ISSUE TITLE
collada2gltf segfault with -k switch (KHR_materials_common)

### ISSUE THREAD

**Issue by timknip** [KhronosGroup/glTF#652]
collada2gltf segfaults on some dae's when using the -k switch.

To reproduce: try with this dae https://dl.dropboxusercontent.com/u/438592/segfault.ZIP
NOTE: the *.bin seems to be written correctly.

Tested with:
1. latest precompiled windows binary
2. collada2gltf compiled from latest source (ubuntu 15)

If needed I can provide a more extensive set of dae's exposing the same problem.

UPDATE:  culprit seems geometry id="geom-ChamferCyl01_e63e11fe04104c95be99137b4ba6c677" in attached file.

**Comment by pjcozzi** [comment]
Thanks for the detailed repo and sample model, @timknip!  Can you workaround this when you change the `id`?

**Comment by timknip** [comment]
@pjcozzi Nope.  Problem persists when using a simpler ID.  
But!
When I remove all other nodes apart from the "problem" node, all is good again. So maybe its simply the amount of geometries being referenced, or the amount of nodes on which -k borks?

**Comment by timknip** [comment]
After some research the issue is clear:

Geometry A - Primitive has VERTEX, NORMAL semantic
Geometry B - Primitive has VERTEX, NORMAL, TEXCOORD semantics

When the primitives of A and B refer to the same material, the segfault occurs.

Here's a python script (deps: pycollada and numpy) to reproduce the segfault:
It'll write out "segfault.dae".

```python
import numpy as np
from collada import Collada
from collada import material,source,geometry,scene
from collada.common import DaeBrokenRefError
from collada.triangleset import BoundTriangleSet

def create_geom(mesh, id, material, use_uv=False):
    vert_floats = [-50,50,50,50,50,50,-50,-50,50,50,-50,50,-50,50,-50,50,50,-50,-50,-50,-50,50,-50,-50]
    normal_floats = [0,0,1,0,0,1,0,0,1,0,0,1,0,1,0,0,1,0,0,1,0,0,1,0,0,-1,0,0,-1,0,0,-1,0,0,-1,0,-1,0,0,-1,0,0,
                    -1,0,0,-1,0,0,1,0,0,1,0,0,1,0,0,1,0,0,0,0,-1,0,0,-1,0,0,-1,0,0,-1]
    tex_floats = np.zeros([len(vert_floats)/3*2], dtype=np.float32)

    vert_src = source.FloatSource("%s-verts-array" % id, np.array(vert_floats), ('X', 'Y', 'Z'))
    normal_src = source.FloatSource("%s-normals-array" % id, np.array(normal_floats), ('X', 'Y', 'Z'))

    geom = geometry.Geometry(mesh, id, id, [vert_src, normal_src])

    input_list = source.InputList()
    input_list.addInput(0, 'VERTEX', "#%s-verts-array" % id)
    input_list.addInput(1, 'NORMAL', "#%s-normals-array" % id)

    indices = np.array([0,0,2,1,3,2,0,0,3,2,1,3,0,4,1,5,5,6,0,
                        4,5,6,4,7,6,8,7,9,3,10,6,8,3,10,2,11,0,12,
                        4,13,6,14,0,12,6,14,2,15,3,16,7,17,5,18,3,
                        16,5,18,1,19,5,20,7,21,6,22,5,20,6,22,4,23])

    if use_uv:
        tex_src = source.FloatSource("%s-texcoords-array" % id, np.array(tex_floats), ('S', 'T'))
        geom = geometry.Geometry(mesh, id, id, [vert_src, normal_src, tex_src])
        tex_indices = np.zeros(indices.size/2, dtype=np.int)
        indices = np.ravel(np.column_stack((indices[0::2], indices[1::2], tex_indices)))
        input_list.addInput(2, 'TEXCOORD', "#%s-texcoords-array" % id)

    triset = geom.createTriangleSet(indices, input_list, material)
    geom.primitives.append(triset)

    return geom

def create():

    mesh = Collada()

    effect = material.Effect("effect0", [], "phong", diffuse=(1,0,0), specular=(0,1,0))
    mat = material.Material("material0", "mymaterial", effect)
    mesh.effects.append(effect)
    mesh.materials.append(mat)

    # NOTE: when both are False, or both are True: no segfault
    # create a geometry without TEXCOORD
    geom = create_geom(mesh, "geometry0", "materialref", False)
    # create a geometry with TEXCOORD
    geom2 = create_geom(mesh, "geometry1","materialref", True)

    mesh.geometries.append(geom)
    mesh.geometries.append(geom2)

    matnode = scene.MaterialNode("materialref", mat, inputs=[])
    matnode2 = scene.MaterialNode("materialref1", mat, inputs=[])

    geomnode = scene.GeometryNode(geom, [matnode])
    node = scene.Node("node0", children=[geomnode])

    geomnode2 = scene.GeometryNode(geom2, [matnode2])
    node2 = scene.Node("node1", children=[geomnode2])

    myscene = scene.Scene("myscene", [node, node2])
    mesh.scenes.append(myscene)
    mesh.scene = myscene

    mesh.write('segfault.dae')

if __name__  == '__main__':
    create()
```

**Comment by pjcozzi** [comment]
Thanks for the script and sample model; this will really help us track it down after we finish the 1.0.1 updates.  Let us know if you track it down before then.

**Comment by timknip** [comment]
@pjcozzi I'll dig into collada2gltf code and try to prepare a PR.

**Comment by pjcozzi** [comment]
That would rock, thanks @timknip!

**Comment by proteamer** [comment]
Hi @timknip, thank you for investigation! I'd like to know about progress on this issue?

**Comment by vt5491** [comment]
FWIW, I found that I always got segmentation faults anytime I had an object that did not have a material (eg. if you just created a plain object).  I am using blender to aframe.  I've found that the best exporter on blender is collada, and the preferred format for A-frame is gltf, thus I've been using collada2gltf.exe a lot lately.

Even so, I would really appreciate any enhancements to prevent this error from occurring, as I seem to get it quite frequently.
